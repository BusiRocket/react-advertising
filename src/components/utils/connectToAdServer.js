import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default PlacementComponent => {
    const HoC = class extends Component {
        constructor(props) {
            super(props);
            this.state = { hidden: false };
        }

        componentDidMount() {
            const { divId } = this.props;
            const { activate } = this.context;
            activate(divId, () => this.setState({ hidden: true }));
        }

        render() {
            if (this.state.hidden) {
                return null;
            }
            return <PlacementComponent {...this.props} />;
        }
    };
    HoC.propTypes = {
        divId: PropTypes.string.isRequired
    };
    HoC.contextTypes = {
        activate: PropTypes.func.isRequired
    };
    return HoC;
};
