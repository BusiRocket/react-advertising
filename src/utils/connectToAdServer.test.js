import { sandbox } from 'sinon';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import connectToAdServer from './connectToAdServer';
import { mount } from 'enzyme';

const mySandbox = sandbox.create();
const mockActivate = mySandbox.spy();
const mockContext = {};

class MockProvider extends Component {
    getChildContext() {
        return {
            activate: (divId, collapseCallback) => {
                mockContext.collapseCallback = collapseCallback;
                return mockActivate(divId, collapseCallback);
            }
        };
    }

    render() {
        return <span id="wrapper">{this.props.children}</span>;
    }
}

MockProvider.childContextTypes = {
    activate: PropTypes.func
};

MockProvider.propTypes = {
    children: PropTypes.node
};

describe('[@mt-advertising/containers/connectToAdServer]', () => {
    describe('When I create a component that is connected to the ad server', () => {
        let MyPlacement;
        beforeEach(() => (MyPlacement = connectToAdServer(({ divId }) => <div id={divId} />)));

        describe('and the component is mounted', () => {
            let wrapper;
            beforeEach(
                () =>
                    (wrapper = mount(
                        <MockProvider>
                            <MyPlacement divId="bla" />
                        </MockProvider>
                    ))
            );
            describe('the render output', () =>
                it('is correct (contains the placement)', () => expect(wrapper.html()).toMatchSnapshot()));
            describe('the “activate” function provided by the wrapper around the component', () => {
                it('is called exactly once', () => void expect(mockActivate).to.have.been.calledOnce);
                it('is called with the correct arguments', () =>
                    void expect(mockActivate.firstCall.args).toMatchSnapshot());
            });
            describe('and I call the collapse callback function passed to the wrapper', () => {
                beforeEach(() => mockContext.collapseCallback());
                describe('the render output', () =>
                    it('is correct (contains not placement, because it is hidden now)', () =>
                        void expect(wrapper.html()).toMatchSnapshot()));
            });
        });
    });
    // https://github.com/sinonjs/sinon/issues/1712
    // Vanilla reset shows a warning!
    afterEach(() => mySandbox.resetHistory());
});
