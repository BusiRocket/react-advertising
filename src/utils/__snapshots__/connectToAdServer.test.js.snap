// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[@mt-advertising/containers/connectToAdServer] When I create a component that is connected to the ad server and the component is mounted and I call the collapse callback function passed to the wrapper the render output is correct (contains not placement, because it is hidden now) 1`] = `"<span id=\\"wrapper\\"><!-- react-empty: 3 --></span>"`;

exports[`[@mt-advertising/containers/connectToAdServer] When I create a component that is connected to the ad server and the component is mounted the render output is correct (contains the placement) 1`] = `"<span id=\\"wrapper\\"><div id=\\"bla\\"></div></span>"`;

exports[`[@mt-advertising/containers/connectToAdServer] When I create a component that is connected to the ad server and the component is mounted the “activate” function provided by the wrapper around the component is called with the correct arguments 1`] = `
Array [
  "bla",
  [Function],
]
`;
